### PR TITLE
Top-level namespace for ToC children entries instead of the assembly name.

### LINKED ISSUE #34557: Assist with Azure SDK partner to update manual TOC part of REF Simplify Python API TOC - Backend changes
Please route to @danieljurek or a point of contact familiar with Python who can address this issue.
Let me know if you have any questions: @KatGreenSmith

Reference API team email: Learn API Docs Team [apidocsteam@microsoft.com](mailto:apidocsteam@microsoft.com)

**Describe the bug**
Contacting Azure SDK partners to make these changes on https://github.com/MicrosoftDocs/azure-docs-sdk-python/blob/main/docs-ref-mapping/reference-unified.yml? [related ADO item: https://dev.azure.com/ceapex/Engineering/_workitems/edit/806157]

We need assistance to replace all distribution package names by lower-level package names. For most packages, it's to replace '-' by '.' on 'children' field. An example is https://github.com/MicrosoftDocs/azure-docs-sdk-python/commit/05303f7b28bd88ba2606692163c1368dd730fd52 

In some packages, the lower-level import package names don't match with distribution names exactly, We suggest to update like below:

See attached Table word doc file for specific details of repo changes. 
[#806157 Simplify Python API TOC - Backend changes.docx](https://github.com/Azure/azure-sdk-for-python/files/14463537/806157.Simplify.Python.API.TOC.-.Backend.changes.docx)

**To Reproduce**
Background:
In azure-docs-sdk-python repo, a [reference-unified.yml](https://github.com/MicrosoftDocs/azure-docs-sdk-python/blob/main/docs-ref-mapping/reference-unified.yml) is used to combine manual written introductions ([example](https://review.learn.microsoft.com/en-us/python/api/overview/azure/eventhub-readme?view=azure-python&branch=main)) with [auto-generated toc](https://github.com/MicrosoftDocs/azure-docs-sdk-python/blob/main/docs-ref-autogen/toc.yml) for OPS build to consume.

Previously in the [reference-unified.yml](https://github.com/MicrosoftDocs/azure-docs-sdk-python/blob/main/docs-ref-mapping/reference-unified.yml), distribution-level package pages are used to concatenate auto-generated toc subtrees. 

As we're removing distribution level package pages in [#806157](https://dev.azure.com/ceapex/d3d54af3-265a-4f18-95f6-9a46397ca583/_workitems/edit/806157), we need to update those distribution-level package pages to lower-level package names.

**Screenshots**
image

**Additional context**

### PR DESCRIPTION
**Pull request by JimSuplizio**
Previously, the ToC for Python MSDocs was using the library name and needed to use the top-level namespace. This is the [workitem](https://dev.azure.com/ceapex/Engineering/_workitems/edit/950573?src=WorkItemMention&src-action=artifact_link) that's driving this work.

**Language-Settings.ps1** - Implemented the function pointer, GetDocsMsDevLanguageSpecificPackageInfoFn, for Python which is called by [Update-DocsMsMetadata.ps1](https://github.com/Azure/azure-sdk-for-python/blob/main/eng/common/scripts/Update-DocsMsMetadata.ps1#L114) when generating/updating the metadata json files in the docs repositories. This function is called after package publish (either official or nightly) and will add the top-level namespace to the json file if it doesn't already exist.

**Docs-ToC.ps1** - Previously this was using the package name in the children entry of the ToC which was wildly incorrect. With these changes it'll use the top-level namespace as the child entry. The functions added and changes are all in support of this.

The [release of azure-template](https://dev.azure.com/azure-sdk/internal/_build/results?buildId=3644579&view=logs&j=19b87903-f8bc-5cbb-22d9-4f6eb3469d8e&t=01c0d070-1636-5aed-84c3-1f03a1861eaf&l=19) was done to test the Language-Settings.ps1 changes which will add the namespace to the json file when it first creates it. This is the ⁠[azure-template.json](https://github.com/MicrosoftDocs/azure-docs-sdk-python/blob/main/metadata/preview/azure-template.json#L15) that's now in the docs repo from that release. Notice that it correctly has its namespace in it.

Fixes #34557 
Fixes azure/azure-sdk-tools#7906

### CODE REVIEW

File: eng/scripts/Language-Settings.ps1 [hunk 1]
```diff
@@ -128,6 +128,34 @@ function Get-python-PackageInfoFromPackageFile ($pkg, $workingDirectory)
...
+        $version = $packageInfo.DevVersion
+      }
+      $namespaces = Get-NamespacesFromWhlFile $packageInfo.Name $version $packageSourceOverride
+      if ($namespaces.Count -gt 0) {
+        $packageInfo | Add-Member -Type NoteProperty -Name "Namespaces" -Value $namespaces
+      } else {
+        LogWarning "Unable to find namespaces for $($packageInfo.Name):$version, using the package name."
+        $tempNamespaces = @()
+        $tempNamespaces += $packageInfo.Name
+        $packageInfo | Add-Member -Type NoteProperty -Name "Namespaces" -Value $tempNamespaces
```
Comment (danieljurek) [inline eng/scripts/Language-Settings.ps1:152]:
```suggestion
$packageInfo | Add-Member -Type NoteProperty -Name "Namespaces" -Value $(packageInfo.Name)
```

File: eng/scripts/docs/Docs-ToC.ps1 [hunk 2]
```diff
@@ -1,4 +1,86 @@
...
+
+  $LibArg= "$Library==$Version"
+  if ($ExtraIndexUrl) {
+    $ExtraIndexArg = "--extra-index-url=$ExtraIndexUrl"
+    Write-Host "python -m pip download --quiet --only-binary=:all: --dest $Destination --no-deps $ExtraIndexArg $LibArg"
+    python -m pip download --quiet --only-binary=:all: --dest $Destination --no-deps $ExtraIndexArg $LibArg
+  } else {
+    Write-Host "python -m pip download --quiet --only-binary=:all: --dest $Destination --no-deps $LibArg"
+    python -m pip download --quiet --only-binary=:all: --dest $Destination --no-deps $LibArg
+  }
```
Comment (danieljurek) [inline eng/scripts/docs/Docs-ToC.ps1:18]: `$PSNativeCommandArgumentPassing = 'Legacy'` is how we've been dealing with this for now. I'm not aware of another pattern for dealing with this at this time. 

No changes needed here at this point.

File: eng/scripts/docs/Docs-ToC.ps1 [hunk 3]
```diff
@@ -1,4 +1,86 @@
...
+        # azure-sdk-for-python repository but there are older libraries that are in the
+        # docs directories which are/were released outside of the repository where this
+        # is true.
+        $rootLevelAzureDir = Join-Path -Path $unpackDir -ChildPath "azure"
+        $namespaceDirs = Get-ChildItem -Path $unpackDir -Recurse -Filter "__init__.py" |
+            Where-Object{$_.DirectoryName -notlike "*.dist-info"} |
+            Where-Object{$_.DirectoryName -notlike "*$([IO.Path]::DirectorySeparatorChar)_*" } |
+            Where-Object{$_.DirectoryName -ine $rootLevelAzureDir}
+        foreach($namespaceDir in $namespaceDirs) {
+            # strip off the root directy, everything left will be subDir1/subDir2 which needs
```
Comment (danieljurek) [inline eng/scripts/docs/Docs-ToC.ps1:64]: Comment incomplete

**Review by danieljurek — APPROVED** [review]
Looks good. Testing also revealed that this works. Before merging we should ensure that the docs team is ready to update the pipeline on their end before Tuesday night's build. 
  
Also, thanks for cleaning up the `$networkService` in `Docs-ToC.ps1`.

File: eng/scripts/Language-Settings.ps1 [hunk 4]
```diff
@@ -618,8 +648,8 @@ function Validate-Python-DocMsPackages ($PackageInfo, $PackageInfos, $PackageSou
 
   $allSucceeded = $true
   foreach ($item in $PackageInfos) {
-    # Some packages 
-    if ($item.Version -eq 'IGNORE') { 
+    # If the Version is INGORE that means it's a source install and those aren't run through ValidatePackage
```
Comment (benbp) [inline eng/scripts/Language-Settings.ps1:651]:
```suggestion
    # If the Version is IGNORE that means it's a source install and those aren't run through ValidatePackage
```

File: eng/scripts/Language-Settings.ps1 [hunk 5]
```diff
@@ -128,6 +128,34 @@ function Get-python-PackageInfoFromPackageFile ($pkg, $workingDirectory)
   }
 }
 
+# This is the GetDocsMsDevLanguageSpecificPackageInfoFn implementation
+function Get-python-DocsMsDevLanguageSpecificPackageInfo($packageInfo, $packageSourceOverride) {
+  # If the default namespace isn't in the package info then it needs to be added
+  # Can't check if (!$packageInfo.Namespaces) in strict mode because Namespaces won't exist
+  # at all.
+  if (!($packageInfo | Get-Member Namespaces)) {
```
[the remark this answers is not in the document]
Comment (JimSuplizio) [reply eng/scripts/Language-Settings.ps1:136]: @benbp this isn't an OR statement. This is doing the following:
```
$foo = $packageInfo | Get-Member Namespaces
if (!$foo) {
}
```

**Pull request merged**